### PR TITLE
Update README.md for #324

### LINKED ISSUE #324: [Feature request]: Explain how to run Fabric with external OpenAI-compatible servers
### What do you need?

One does not need to pay for ChatGPT while there are many free inference servers around. For example, academia can (and does) deploy their own models in openai-compatible servers. I am writing a pull request to add this to the documentation. I already tested it and it works.

### PR DESCRIPTION
**Pull request by surak**
## What this Pull Request (PR) does
It's just an update to README, showing how to point to your own server.

## Related issues
Closes #324 , showing how to connect to another server

### CODE REVIEW

**Comment by ksylvan** [comment]
This doesn't currently work with the current code, till #334 is merged. This is because the way the `OpenAi()` client is instantiated we don't pass the base URL.

File: README.md [hunk 1]
```diff
@@ -203,6 +203,15 @@ fabric --help
 
 ### Using the `fabric` client
 
+If you want to use it with OpenAI API compatible inference servers, such as [FastChat](https://github.com/lm-sys/FastChat) or [Helmholtz Blablador](http://helmholtz-blablador.fz-juelich.de), simply export the following environment variables:
+
+- `export OPENAI_BASE_URL=https://YOUR-SERVER:8000/v1/`
```
Comment (ksylvan) [inline README.md:208]: Change the environment variable to `OPENAI_API_BASE_URL` which I've seen used elsewhere and is used in #334 

Also add "LM Studio" at https://lmstudio.ai/ to the list of OpenAI compatible servers.

**Review by ksylvan — CHANGES_REQUESTED** [review]
Please change the

File: README.md (hunk 1 above)
Comment (surak) [reply README.md:208]: Hi Kayvan, 

The variable is `OPENAI_BASE_URL`, without the `API` : 
https://github.com/openai/openai-python/blob/0cbce3f6777fd8f137ed7445e9bc2f2aa5c9f5b4/src/openai/_client.py#L110

As you can see, there is no mention to this name on their code at all: 
https://github.com/search?q=repo%3Aopenai%2Fopenai-python%20OPENAI_API_BASE_URL&type=code

Comment (ksylvan) [reply README.md:208]: Awesome. My PR is unnecessary, then. I'll test and close it. Thank you.

**Review by ksylvan — APPROVED** [review]
Looks good to me. Tested it with LM Studio.

**Comment by scaruslooner** [comment]
Does this mean this works with Groq as well. (https://console.groq.com/docs/openai)

**Comment by ksylvan** [comment]
> Does this mean this works with Groq as well. (https://console.groq.com/docs/openai)

@scaruslooner I don't see why it would not. You'd just need to know the URL for the Groq API.

**Comment by ksylvan** [comment]
@danielmiessler @xssdoctor can we get this simple documentation change into the repo?

**Comment by scaruslooner** [comment]
how do you change the url on windows. I tried using set OPENAI_API_BASE_URL=https://api.groq.com/openai/v1. But got an error: 
(fabric) C:\Users\camer\Documents\fabric>set OPENAI_API_BASE_URL=https://api.groq.com/openai/v1
(fabric) C:\Users\camer\Documents\fabric>
(fabric) C:\Users\camer\Documents\fabric>fabric --pattern summarize
Error: Client error '401 Unauthorized' for url 'https://api.openai.com/v1/models'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401

**Comment by ksylvan** [comment]
You need to also set OPENAI_API_KEY to your Groq API key otherwise it will try to use your OpenAI API key from your `.env` file.

On Mac, I did this to run the one command with modified environment. My Groq API key was set in `$groq_api`

```bash
% env OPENAI_BASE_URL=https://api.groq.com/openai/v1 OPENAI_API_KEY=$groq_api fabric --pattern ai -m Llama2-70b-4096 --text 'Discuss the historical significance of the Silk Road in global trade.'
The Silk Road, a network of ancient trade routes connecting China and the Far East with the Middle East and Europe, was established during the Han dynasty (206 BCE - 220 CE) and remained an important conduit for global trade until the 15th century. The Silk Road was historically significant for several reasons:
[...]
```

PowerShell example:

```powershell
PS C:\Users\kayvan> $env:OPENAI_BASE_URL = "https://api.groq.com/openai/v1"
PS C:\Users\kayvan> $env:OPENAI_API_KEY="gsk_REDACTED"
PS C:\Users\kayvan> fabric --pattern ai -m Llama2-70b-4096 --text 'Discuss the historical significance of the Silk Road in global trade.'
The Silk Road, a network of ancient trade routes connecting China and the Far East with the Middle East and Europe, was established during the Han dynasty (206 BCE - 220 CE) and remained an important conduit for global trade until the 15th century. The Silk Road was historically significant for several reasons:
1. Facilitated the exchange of goods: The Silk Road facilitated the exchange of goods such as silk, spices, tea, precious stones, and metals between the East and the West.
[...]
```

On CMD shell, you can do this:

```cmd
C:\Users\kayvan>set OPENAI_API_KEY=gsk_REDACTED
C:\Users\kayvan>set OPENAI_BASE_URL=https://api.groq.com/openai/v1
C:\Users\kayvan>fabric --pattern ai -m Llama2-70b-4096 --text "Discuss the historical significance of the Silk Road in global trade."
The Silk Road, a network of ancient trade routes connecting China and the Far East with the Middle East and Europe, was established during the Han dynasty (206 BCE - 220 CE) and remained an important conduit for global trade until the 15th century. The Silk Road was historically significant for several reasons:
1. Facilitated the exchange of goods: The Silk Road facilitated the exchange of goods such as silk, spices, tea, precious stones, and metals between the East and the West.
[...]
```

**Comment by scaruslooner** [comment]
Thank you so much :) This work!

File: README.md [hunk 2]
```diff
@@ -203,6 +203,15 @@ fabric --help
...
 
+If you want to use it with OpenAI API compatible inference servers, such as [FastChat](https://github.com/lm-sys/FastChat), [Helmholtz Blablador](http://helmholtz-blablador.fz-juelich.de), [LM Studio](https://lmstudio.ai) and others, simply export the following environment variables:
+
+- `export OPENAI_BASE_URL=https://YOUR-SERVER:8000/v1/`
+- `export DEFAULT_MODEL="YOUR_MODEL"`
+
+And if your server needs authentication tokens, like Blablador does, you export the token the same way you would with OpenAI:
+  
+- `export OPENAI_API_KEY="YOUR TOKEN"`
+
```
Comment (camilosampedro) [inline README.md:214]: I tried exporting these variables, before running fabric, setting them to point to my local LM Studio server, and they only seem to be working with the `--gui`:

```shell script
export OPENAI_BASE_URL=http://localhost:1234/v1
export DEFAULT_MODEL="QuantFactory/Meta-Llama-3-8B-Instruct-GGUF"
export OPENAI_API_KEY=""

pbpaste | fabric --pattern improve_writing --stream
```

Errors with:
```
Error: Connection error.
Connection error.
```

But when I set the same environment variables and run with the gui parameter, it seems to start working:
```shell script
fabric --gui
# Then use the gui to input text
```

🤔

Comment (ksylvan) [reply README.md:214]: It works with both `--gui` and commandline for me with LM studio server. But I'm not setting the OPENAI_API_KEY to `""`

Comment (ksylvan) [reply README.md:214]: @camilosampedro 

Running the new small phi-3 Microsoft model via LM Studio

```text
env OPENAI_BASE_URL=http://localhost:1234/v1 fabric --listmodels
GPT Models:
Phi-3-mini-4k-instruct-fp16.gguf

Local Models:
llama2:latest
llama3:latest

Claude Models:
claude-3-opus-20240229
claude-3-sonnet-20240229
claude-3-haiku-20240307
claude-2.1
```

And now I did this:

```text
% env OPENAI_BASE_URL=http://localhost:1234/v1 fabric -m Phi-3-mini-4k-instruct-fp16.gguf --pattern raw_query
Answer these 2 questions:
1) Why is the sky blue?
2) What is the tallest tree in the world?
The color of the sky appears blue due to a phenomenon called Rayleigh scattering. As sunlight enters Earth's atmosphere, it collides with molecules and small particles present therein. These interactions cause shorter wavelength colors (blue and violet) to scatter more than longer wavelength colors (red and yellow). However, our eyes are more sensitive to blue light, which is why we perceive the sky as predominantly blue during daylight hours.

The tallest tree in the world is a coast redwood known as Hyperion. It stands at an impressive height of 379.7 feet (115.7 meters) and was discovered on August 25, 2006, by naturalists Chris Atkins and Michael Taylor.
```

But when I set OPENAI_API_KEY to blank like in your example, I get exactly the same error:

```text
% env OPENAI_API_KEY="" OPENAI_BASE_URL=http://localhost:1234/v1 fabric -m Phi-3-mini-4k-instruct-fp16.gguf --pattern raw_query
Make me a haiku about Large Language Models.
Error: Connection error.
Connection error.
```

Try it again without setting OPENAI_API_KEY to "" or set it to some random value.

```text
% env OPENAI_API_KEY="N/A" OPENAI_BASE_URL=http://localhost:1234/v1 fabric -m Phi-3-mini-4k-instruct-fp16.gguf --pattern raw_query
Make me a haiku about Large Language Models.
Silent minds awake,
Words flow like rivers in code,
AI's vast knowledge blooms.
```

**Pull request merged**